ibis/expr/timecontext.py

```diff
@@ -0,0 +1,190 @@
+    left_context: TimeContext
+    right_context: TimeContext
+
+    Returns
+    -------
+    result : TimeContextRelation
+    """
+    left_begin, left_end = left_context
+    right_begin, right_end = right_context
+    if right_begin <= left_begin and right_end >= left_end:
+        return TimeContextRelation.SUBSET
+    elif right_begin >= left_begin and right_end <= left_end:
+        return TimeContextRelation.SUPERSET
+    elif right_end < left_begin or left_end < right_begin:
+        return TimeContextRelation.NONOVERLAP
+    else:
+        return TimeContextRelation.OVERLAP
+
+
+def canonicalize_context(
+    timecontext: Optional[TimeContext],
+) -> Optional[TimeContext]:
+    """Convert a timecontext to canonical one with type pandas.Timestamp
+       for its begin and end time. Raise Exception for illegal inputs
+    """
+    SUPPORTS_TIMESTAMP_TYPE = pd.Timestamp
+    try:
+        begin, end = timecontext
+    except (ValueError, TypeError):
+        raise com.IbisError(
+            f'Timecontext {timecontext} should specify (begin, end)'
+        )
+
+    if not isinstance(begin, SUPPORTS_TIMESTAMP_TYPE):
+        raise com.IbisError(
+            f'begin time value {begin} of type {type(begin)} is not'
+            ' of type pd.Timestamp'
+        )
+    if not isinstance(end, SUPPORTS_TIMESTAMP_TYPE):
+        raise com.IbisError(
+            f'end time value {end} of type {type(begin)} is not'
+            ' of type pd.Timestamp'
+        )
+    if begin > end:
+        raise com.IbisError(
+            f'begin time {begin} must be before or equal' f' to end time {end}'
+        )
+    return begin, end
+
+
+""" Time context adjustment algorithm
+    In an Ibis tree, time context is local for each node, and they should be
+    adjusted accordingly for some specific nodes. Those operations may
+    require extra data outside of the global time context that user defines.
+    For example, in asof_join, we need to look back extra `tolerance` daays
+    for the right table to get the data for joining. Similarly for window
+    operation with preceeding and following.
+    Algorithm to calculate context adjustment are defined in this module
+    and could be used by multiple backends.
+"""
+
+
+@functools.singledispatch
+def adjust_context(op, *clients, timecontext: TimeContext) -> TimeContext:
+    """
+    Params
+    -------
+    op: ibis.expr.operations.Node
+    clients: List[ibis.client.Client], backend for execution
+    timecontext: TimeContext, time context associated with the node
+
+    Returns
+    --------
+    Adjusted time context
+    """
+    raise NotImplementedError()
+
+
+@adjust_context.register(ops.AsOfJoin)
+def adjust_context_asof_join(
+    op, *clients, timecontext: TimeContext
+) -> TimeContext:
+    (backend,) = clients
+    begin, end = timecontext
+
+    if op.tolerance is not None:
+        timedelta = backend.execute(op.tolerance)
+        # only backwards and adjust begin time only
```

Comment:
Can you raise for forward looking window?

Reply:
I mean, forward looking join.

Reply:
+1

Reply:
Actually there is no way to specify forward-looking here, since in ibis AsOfJoin node there is no 'direction' arg, it is always looking backward.

Reply:
kk. Looks good then.